### PR TITLE
Add Compact Convolutional Transformer model (CCT)

### LINKED ISSUE #20133: Compact Transformer
### Model description

# Escaping the Big Data Paradigm with Compact Transformers

Abstract :

> With the rise of Transformers as the standard for language processing, and their advancements in computer vision, there has been a corresponding growth in parameter size and amounts of training data. Many have come to believe that because of this, transformers are not suitable for small sets of data. This trend leads to concerns such as: limited availability of data in certain scientific domains and the exclusion of those with limited resource from research in the field. In this paper, we aim to present an approach for small-scale learning by introducing Compact Transformers. We show for the first time that with the right size, convolutional tokenization, transformers can avoid overfitting and outperform state-of-the-art CNNs on small datasets. Our models are flexible in terms of model size, and can have as little as 0.28M parameters while achieving competitive results. Our best model can reach 98% accuracy when training from scratch on CIFAR-10 with only 3.7M parameters, which is a significant improvement in data-efficiency over previous Transformer based models being over 10x smaller than other transformers and is 15% the size of ResNet50 while achieving similar performance. CCT also outperforms many modern CNN based approaches, and even some recent NAS-based approaches. Additionally, we obtain a new SOTA result on Flowers-102 with 99.76% top-1 accuracy, and improve upon the existing baseline on ImageNet (82.71% accuracy with 29% as many parameters as ViT), as well as NLP tasks. Our simple and compact design for
transformers makes them more feasible to study for those with limited computing resources and/or dealing with small
datasets, while extending existing research efforts in data efficient transformers.

### Open source status

- [X] The model implementation is available
- [X] The model weights are available

### Provide useful links for the implementation

Paper : https://arxiv.org/pdf/2104.05704.pdf
Github repository : https://github.com/SHI-Labs/Compact-Transformers

### PR DESCRIPTION
**Pull request by rishabbala**
# What does this PR do?

Fixes #20133 (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@amyeroberts @ArthurZucker

### CODE REVIEW

**Comment by rishabbala** [comment]
Could you let me know how to fix the failed test. The issue apparently is that PIL.Image.LINEAR does not exist

**Comment by ArthurZucker** [comment]
Rebase on main, a fix has been merge! 😉

**Comment by rishabbala** [comment]
Done. Could you let me know if there are any further changes

**Comment by ArthurZucker** [comment]
Hey! Sorry I forgot to mention this in my previous comment, we are trying to push models to the hub as often as possible, to make it a LOT easier for contributors! Then we share the repo on the hub where everything is supported 🤗 
I would recommend following [this tutorial](https://huggingface.co/docs/transformers/custom_models), and sharing here the uploaded model! Tell me if that sounds good to you!

**Comment by rishabbala** [comment]
Sorry if I misunderstood, but I followed this [tutorial](https://huggingface.co/docs/transformers/add_new_model#514-port-brandnewbert-to-transformers), and it looks similar to the tutorial you shared. Could you quickly point out the difference between the two, or what additionally I must do?

**Comment by ArthurZucker** [comment]
Sure, adding the model to the hub rather than with a PR to transformers would give a model usable out of the box, without having to fix all of the CI, while keeping your code and no need for reviews etc. Mostly you would have to add MAPPINGS as explain in the tutorial, and users will just need to use `trust_remote_code = True` when doing `class.from_pretrained`!

**Comment by rishabbala** [comment]
I've made the changes, and uploaded the models: [cct_224](https://huggingface.co/rishabbala/cct_14_7x2_224) [cct_384](https://huggingface.co/rishabbala/cct_14_7x2_384). Let me know if this looks ok, and if I should go ahead and close the PR. Thanks for the help :)

**Comment by ArthurZucker** [comment]
Look good to me thanks a lot! Would suggest you to add a model card for people who don't really know the model and might want a quick way to use it!
